planner/core/rule_partition_processor.go

```diff
@@ -224,13 +224,94 @@ func (s *partitionProcessor) processHashPartition(ds *DataSource, pi *model.Part
 	return tableDual, nil
 }
 
-func (s *partitionProcessor) pruneListPartition(tbl table.Table, partitionNames []model.CIStr) []int {
-	if len(partitionNames) > 0 {
-		pi := tbl.Meta().Partition
-		or := partitionRangeOR{partitionRange{0, len(pi.Definitions)}}
-		return s.convertToIntSlice(or, pi, partitionNames)
+func (s *partitionProcessor) findUsedListPartitions(ctx sessionctx.Context, tbl table.Table, partitionNames []model.CIStr,
+	conds []expression.Expression) ([]int, error) {
+	pi := tbl.Meta().Partition
+	partExpr, err := tbl.(partitionTable).PartitionExpr()
+	if err != nil {
+		return nil, err
+	}
+
+	colsIDToUniqueID := make(map[int64]int64)
+	for _, cond := range conds {
+		condCols := expression.ExtractColumns(cond)
+		for _, c := range condCols {
+			colsIDToUniqueID[c.ID] = c.UniqueID
```

Comment:
Since the `col.UniqueID` may be different in the `union` or `join` statement, So I use the `col.ID` to validate the same column. @lysu @tiancaiamao @XuHuaiyu PTAL